### PR TITLE
Feature/fix issues from security contest 20230517

### PR DESCRIPTION
**Pull request by ksin751119**
- For loop variable doesn't need to set to zero
- The opposite meaning isPaused modifier is whenNotPaused
- WrappedNative doesn't need to convert to address type
- Replace the `address to` with `address` if it is unused in calculator.getFees()
- Set daiToken of MakerDrawFeeCalculator to immutable
- add _newAgent(address user) internal function to save gas cost in execute/executeWithSignature
- Unnecessary uint160 casting when calculating agent address
- Rename  getUserAgent() to getCurrentUserAgent()

### CODE REVIEW

**Comment by zodahu** [comment]
LGTM

**Comment by zodahu** [comment]
There appear to be 18 instances of `uint256 i =` in `src/`, could you help check?

**Review by JavanC — COMMENTED** [review]
- resume or unpause like Pausable.sol?
- the modifier in pause/unpause function is need?

File: src/Router.sol [hunk 1]
```diff
@@ -225,13 +225,17 @@ contract Router is IRouter, EIP712, FeeGenerator {
...
-            return payable(address(agent));
+            return _newAgent(user);
         }
     }
 
+    function _newAgent(address user) internal returns (address payable) {
+        IAgent agent = IAgent(address(new Agent{salt: bytes32(bytes20(user))}(agentImplementation)));
+        agents[user] = agent;
+        emit AgentCreated(address(agent), user);
+        return payable(address(agent));
```
Comment (zodahu) [inline src/Router.sol:236]:
```suggestion
    function _newAgent(address user) internal returns (address) {
        IAgent agent = IAgent(address(new Agent{salt: bytes32(bytes20(user))}(agentImplementation)));
        agents[user] = agent;
        emit AgentCreated(address(agent), user);
        return address(agent);
```

File: src/fees/FeeGenerator.sol [hunk 2]
```diff
@@ -122,13 +96,39 @@ abstract contract FeeGenerator is Ownable {
...
         IParam.Fee[] memory fees = new IParam.Fee[](realFeeLength);
-        for (uint256 i = 0; i < realFeeLength; ++i) {
+        for (uint256 i; i < realFeeLength; ++i) {
             fees[i] = tempFees[i];
         }
 
         return fees;
     }
 
+    /// @notice Set fee calculator contracts
```
Comment (zodahu) [inline src/fees/FeeGenerator.sol:106]: Jeff's [PR](https://github.com/dinngo/composable-router-contract/pull/70/files#diff-4c5c1b304f84735526561571301bebb623d96d819fe3f19fd7b294cde2df7d79) has move the two functions upward.

File: src/Router.sol [hunk 3]
```diff
@@ -151,10 +151,10 @@ contract Router is IRouter, EIP712, FeeGenerator {
         emit Paused();
     }
 
-    /// @notice Resume `execute` and `executeWithSignature` by pauser
-    function resume() external onlyPauser {
+    /// @notice Unpause `execute` and `executeWithSignature` by pauser
+    function unpause() external onlyPauser {
         paused = false;
```
Comment (Jeff-CCH) [inline src/Router.sol:156]: Wondering if `delete paused` is better?

File: src/fees/FeeGenerator.sol (hunk 2 above)
Comment (ksin751119) [reply src/fees/FeeGenerator.sol:106]: Sorry, it must be the wrong adjustment when rebasing PR

File: src/Router.sol [hunk 4]
```diff
@@ -196,7 +196,7 @@ contract Router is IRouter, EIP712, FeeGenerator {
         IParam.Logic[] calldata logics,
         address[] calldata tokensReturn,
         uint256 referralCode
-    ) external payable isPaused checkCaller {
+    ) external payable whenNotPaused checkCaller {
```
[the remark this answers is not in the document]
Comment (ksin751119) [reply src/Router.sol:169]: Just follow openzeppelin [Pausable](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/security/Pausable.sol).sol naming from Javan suggestion

File: src/Router.sol (hunk 3 above)
Comment (ksin751119) [reply src/Router.sol:156]: I think most cases to use delete variables are struct variables. For me, using assign false here is more suitable.

File: src/fees/FeeGenerator.sol (hunk 2 above)
Comment (zodahu) [reply src/fees/FeeGenerator.sol]: Still noticing a difference in `FeeGenerator.sol` file changed, the two functions should be moved to line 37.

Comment (ksin751119) [reply src/fees/FeeGenerator.sol]: Sorry, Fixed it.

**Comment by zodahu** [comment]
LGTM

File: src/Router.sol (hunk 3 above)
Comment (Jeff-CCH) [reply src/Router.sol:156]: hmm, I'm not sure if it's still the case that we can get a gas refund from `delete`. If it's not the case anymore, either way is fine.

Comment (ksin751119) [reply src/Router.sol:156]: I like to maintain the original approach; it's uncommon to come across other protocols using the "delete" method to reset a uint256 variable. Even OpenZeppelin assigns "false" to the "pause" variable in their `Pausable.sol` implementation.

**Comment by Jeff-CCH** [comment]
LGTM

File: src/Router.sol (hunk 3 above)
Comment (zodahu) [reply src/Router.sol:156]: There is no difference between `delete` and false assignment on gas refund.

**Pull request merged**